### PR TITLE
don't recheck events in `select_from_last_begin`

### PR DESCRIPTION
**Pull request by ces42**
This addresses part of https://github.com/iberianpig/fusuma/issues/295 . It fixes the inefficiency in https://github.com/iberianpig/fusuma/blob/b68ab60335260dc32bfb3748c0bf65f9dc6d9dad/lib/fusuma/plugin/buffers/gesture_buffer.rb#L99-L107
`.reverse.find` has bad runtime because it creates a reversed copy of the list (in linear time) and then iterates over it. But of course one should just iterate backwards over the list. 

Furthermore instead of searching the whole buffer at every call of `select_from_last_begin` we should just search all entries that have been added since the last call and if none of them are a "begin" we just return the same as last time. This is what the two new variables `@mem_last_begin` and `@mem_checked` are for. 
When I run fusuma, the "begin" event is always at index 0 because the buffer is `clear`ed before the next gesture starts and there's always one more entry in `@events` than in the last call. So this optimization is very efficient.

Finally, if the last "begin" did happen at index 0 we can return `self` instead of returning an identical copy.

**possible problems that this could cause:**
Things will break if
- `@events` is modified in-place in any way other than appending events (e.g. `pop()`)
- `select_from_last_begin` returns `self` and that return value is modified

**profiling**
I enabled the line profiler and ran `master` vs my fork while (3-finger) swiping from one end of my touchpad to the other end (separate swipes left, right, left, right, etc.) with `time exe/fusuma`

current master:
```
lib/fusuma/plugin/detectors/swipe_detector.rb                                                                                                                                 
               |  18          # @return [NilClass] if event is NOT detected
               |  19          def detect(buffers)
   3.4ms  2682 |  20            gesture_buffer = buffers.find { |b| b.type == BUFFER_TYPE }
 315.3ms   894 |  21              .select_from_last_begin
 462.7ms 302172 |  22              .select_by_events { |e| e.record.gesture == GESTURE_RECORD_TYPE }
               |  23  
 310.4ms   894 |  24            updating_events = gesture_buffer.updating_events
   0.4ms   894 |  25            return if updating_events.empty?
```

this fork:
```
lib/fusuma/plugin/detectors/swipe_detector.rb
               |  18          # @return [NilClass] if event is NOT detected
               |  19          def detect(buffers)
   3.2ms  2745 |  20            gesture_buffer = buffers.find { |b| b.type == BUFFER_TYPE }
   9.3ms   915 |  21              .select_from_last_begin
 654.5ms 449571 |  22              .select_by_events { |e| e.record.gesture == GESTURE_RECORD_TYPE }
               |  23  
 454.9ms   915 |  24            updating_events = gesture_buffer.updating_events
   0.5ms   915 |  25            return if updating_events.empty?
```

### CODE REVIEW

File: lib/fusuma/plugin/filters/libinput_device_filter.rb [hunk 1]
```diff
@@ -57,13 +58,18 @@ def initialize(name_patterns:)
...
           end
 
+          def filter_record(record)
+            @accept_any || all.any? { |device| record.to_s =~ /^[\s-]?#{device.id}\s/ }
+          end
+
           # @return [Array]
           def all
             @all ||= if @name_patterns.empty?
+              @accept_any = true
```
Comment (iberianpig) [inline lib/fusuma/plugin/filters/libinput_device_filter.rb:72]: It appears that the use of `@accept_any = true` is causing the filtering by `Device.available` to be ineffective.

**Review by iberianpig — CHANGES_REQUESTED** [review]

**Comment by iberianpig** [comment]
@ces42 
I apologize for the delayed review of the pull request.
Thank you for tuning the performance.
I have made some comments.

File: lib/fusuma/plugin/filters/libinput_device_filter.rb (hunk 1 above)
Comment (ces42) [reply lib/fusuma/plugin/filters/libinput_device_filter.rb:72]: I see. Is this an issue? I assumed that the use of filtering would be "negative filtering", i.e. something like "don't make fusuma try to process my touchscreen events as gestures". It seems that if there's no `name_patterns` then there is no need to filter events? I also expected that events would always come from "available" devices, is that wrong?

**Comment by ces42** [comment]
Hi! I'm traveling right now so I probably won't have time to look at this until next week. I have made some more changes (locally) and I split the ones here into three separate branches, maybe that will make reviewing easier.

**Pull request closed without merging**